Ready: Time Series API

alexmoore opened the pull request:
Finally ready. Contains all the new classes/commands/operations for the new TS API. 
Also contains some general code formatting cleanup, as well as removing support for Java 6.

alexmoore committed:
Implementing time series Query operation.

alexmoore committed:
Adding Query command & param checking.

alexmoore committed:
First pass at TS Store operation

alexmoore committed:
Adding integration test for Store operation.

alexmoore committed:
Reworking Cell class to be more fluid and happy.

alexmoore committed:
Adding TS Put command & Tests"

alexmoore committed:
Adding some conversion tests for Cell class

alexmoore committed:
Update interpolation validity regex

alexmoore committed:
Comment cleanup

alexmoore committed:
Cleanup, adding new SetCell helper interface.

alexmoore committed:
Fix test equality

alexmoore committed:
Adding "map" helper class

alexmoore committed:
Fixing numerics to use Binary coded decimal

alexmoore committed:
fix test

srgg commented on src/main/java/com/basho/riak/client/api/commands/timeseries/Query.java:
```diff
@@ -0,0 +1,112 @@
+public class Query extends RiakCommand<QueryResult, BinaryValue>
+{
+    private final Builder builder;
+    private final Logger logger = LoggerFactory.getLogger(Query.class);
```
better to declare as static

alexmoore replied:
The logger?

srgg replied:
yep

srgg commented on src/main/java/com/basho/riak/client/api/commands/timeseries/Query.java:
```diff
@@ -0,0 +1,112 @@
+    {
+        private final Logger logger = LoggerFactory.getLogger(Query.Builder.class);
+
+        private final Pattern paramPattern = Pattern.compile("(:[a-zA-Z][0-9a-zA-Z_]*)");
```
better to declare as static

srgg commented on src/main/java/com/basho/riak/client/core/converters/TimeSeriesConverter.java:
```diff
@@ -0,0 +1,203 @@
+
+        for (RiakKvPB.TsRow pbRow : pbRows)
+        {
+            List<Cell> cells = new ArrayList<Cell>();
```
Better to create a list of required size, if possible:

 new ArrayList<Cell>(pbRow.getCellsList().size()); // No sure about availability of this method

srgg commented on src/main/java/com/basho/riak/client/core/converters/TimeSeriesConverter.java:
```diff
@@ -0,0 +1,203 @@
+{
+    private List<Row> parseRows(List<RiakKvPB.TsRow> pbRows)
+    {
+        ArrayList<Row> rows = new ArrayList<Row>();
```
Better to create a list of required size: 

new ArrayList<Row>(pbRows.size());

alexmoore committed:
Make some objects static

srgg commented on src/main/java/com/basho/riak/client/core/converters/TimeSeriesConverter.java:
```diff
@@ -0,0 +1,203 @@
+            {
+                if (pbCell.hasBinaryValue())
+                {
+                    cells.add(new Cell(BinaryValue.unsafeCreate(pbCell.getBinaryValue().toByteArray())));
```
For better readability it may be useful to introduce a factory method:

Cell.create()

And maybe move all the magic inside it

hazen replied:
Agreed. That is a pretty verbose line already.

alexmoore replied:
But then our Cell knows about every network protocol implementation.

srgg replied:
You are totally right about this case, but at the moment, since we have the only one protocol from my point of view it's better to introduce template method. Later, if that become a case, we may do simple refactoring to introduce AbstractFactory with PBCellConcreteFactory and WhateEverProtoConcreteFactory..

if we are going to move magic inside the factory method, better to call the method kinda fromPB(..)

alexmoore committed:
Code review cleanup, staticizing loggers, finalizing members, initializing arrays

srgg commented on src/main/java/com/basho/riak/client/core/converters/TimeSeriesConverter.java:
```diff
@@ -0,0 +1,203 @@
+ * @author Alex Moore <amoore at basho dot com>
+ * @since 2.0.3
+ */
+public class TimeSeriesConverter
```
To reduce memory consumption, since it is stateless and without polymorphic behavior, better to use as a kind of static facade.

srgg commented on src/main/java/com/basho/riak/client/core/operations/TimeSeriesQueryOperation.java:
```diff
@@ -0,0 +1,130 @@
+        // This is not a streaming op, there will only be one response
+        if (responses.size() > 1)
+        {
+            logger.error("Received {} responses when only one was expected.", responses.size());
```
to manage log reader's expectations it is better to mention what will happen with the extra responses

srgg committed:
introduce generic PB Future operation

alexmoore committed:
Staticize PBConverter class

alexmoore committed:
Renaming some methods to be more descriptive

alexmoore committed:
Merge branch 'develop' into timeseries

srgg commented on src/main/java/com/basho/riak/client/api/commands/timeseries/Query.java:
```diff
@@ -0,0 +1,114 @@
+
+            if(!paramMatcher.matches()) { return; }
+
+            for (int i = 0; i < paramMatcher.groupCount(); i++) {
```
it's better to instantiate knownParams here, since we know exact size

srgg commented on src/main/java/com/basho/riak/client/api/commands/timeseries/Query.java:
```diff
@@ -0,0 +1,114 @@
+            Matcher paramMatcher = paramPattern.matcher(queryText);
+
+            if(!paramMatcher.matches()) { return; }
+
```
it's better to instantiate knownParams here, since we know exact size

srgg commented on src/main/java/com/basho/riak/client/api/commands/timeseries/Query.java:
```diff
@@ -0,0 +1,114 @@
+
+        public Builder addStringParameter(String key, String value)
+        {
+            checkParamValidity(key);
```
better to move checkParamValidity(key);] call to the  private Builder addParameter(BinaryValue key, BinaryValue value)

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/Cell.java:
```diff
@@ -0,0 +1,227 @@
+
+    public boolean hasLong()
+    {
+        return this.isIntegerCell;
```
The range of valid Long  is wider than Integer

alexmoore replied:
`this.isIntegerCell` is a boolean that signifies that something has been put into the integerValue register, I think you're thinking of hasInt(). There's an impedance mismatch between Java Types and TS types.

srgg replied:
~~But, by using this method https://github.com/basho/riak-java-client/pull/543/files#diff-4e432a46681bf2317f99d9c32411a94dR37 cell might be created with real java Long (without truncation https://github.com/basho/riak-java-client/pull/543/files#diff-4e432a46681bf2317f99d9c32411a94dR74) and for this cell the both methods hasInt() and  hasLong() returns false~~

Ohhhh, I see yes you are right I was thinking about hasInt()

alexmoore replied:
Yeah, Time Series "Integer" == "64-bit signed integer" == Java Long.  Wish we could figure out some better names.

srgg replied:
gotcha, thank you for the details

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/Cell.java:
```diff
@@ -0,0 +1,227 @@
+    protected byte[][] setValue;
+    protected byte[] mapValue;
+
+    private boolean isIntegerCell = false;
```
To reduce memory consumption and increase readability it might be better to introduce CellType enum (or re-use the existing ColumnType). As a result we also may decide to get rid of hasXValue() methods. And that will allow to handle cell types by using switch instead of cooking if (x.hasXvalue)

alexmoore replied:
Agreed, but this may go away with a Cell factory.

srgg replied:
But even with the factory it will be nice to have kind of

``` java

    public Cell(int value){
      //..
      cellType = ColumnType. INTEGER;
    }

    public  ColumnType cellType(){
      return cellType;
    }
```

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/MapCell.java:
```diff
@@ -0,0 +1,48 @@
+    @SuppressWarnings("unchecked")
+    public static <T> T getObject(Cell cell, TypeReference<T> typeReference)
+    {
+        T obj;
```
more natural to use blank final here

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/ColumnDescription.java:
```diff
@@ -0,0 +1,75 @@
+ * @author Alex Moore <amoore at basho dot com>
+ * @since 2.0.3
+ */
+public class ColumnDescription
```
it might be a good idea to add proper equals() and hashCode()  methods as well as toString()

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/SetCell.java:
```diff
@@ -0,0 +1,59 @@
+
+    public static <T> SetCell fromSet(Set<T> value, TypeReference<T> typeReference)
+    {
+        SetCell cell = new SetCell();
```
it's better to re-use Cell.newSet

alexmoore replied:
But then we return a Cell, not a SetCell.  Again, may be reworked with the factories.

srgg replied:
As an option - We may change the signature of https://github.com/basho/riak-java-client/pull/543/files#diff-4e432a46681bf2317f99d9c32411a94dR212 to return SetCell

srgg commented on src/test/java/com/basho/riak/client/api/commands/itest/ITestTimeSeries.java:
```diff
@@ -0,0 +1,111 @@
+        }
+        catch (ExecutionException ex)
+        {
+            System.out.println(ex.getCause().getCause());
```
better to use logger

alexmoore replied:
Fixed.

lukebakken commented on src/test/java/com/basho/riak/client/api/commands/itest/ITestTimeSeries.java:
```diff
@@ -0,0 +1,111 @@
+        final Calendar date1 = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        date1.add(Calendar.MILLISECOND, -10);
+        final Calendar date2 = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        date1.add(Calendar.MILLISECOND, -5);
```
Should be `date2.add...` here

lukebakken commented on src/test/java/com/basho/riak/client/core/operations/itest/ITestTimeSeriesStoreOperations.java:
```diff
@@ -0,0 +1,57 @@
+        final Calendar date1 = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        date1.add(Calendar.MILLISECOND, -10);
+        final Calendar date2 = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        date1.add(Calendar.MILLISECOND, -5);
```
Should be `date2.add...` here

alexmoore replied:
Fixed.

alexmoore committed:
Merge branch 'develop' into timeseries

alexmoore committed:
Fixes from PR review

srgg commented on src/main/java/com/basho/riak/client/api/commands/timeseries/Query.java:
```diff
@@ -0,0 +1,122 @@
+
+            if(!paramMatcher.matches())
+            {
+                knownParams = new HashSet<String>(0);
```
better to use immutable java.util.Collections.EMPTY_SET

alexmoore committed:
Using static empty set instead of new HashSet(0)

alexmoore committed:
Updating client to use new interpolation API

alexmoore committed:
Updating Tests to new syntax.

alexmoore committed:
Revert "Updating Tests to new syntax."

This reverts commit b70e97a6e12e05df25d0f0dd0f87d31293384ae8.

alexmoore committed:
Revert "Updating client to use new interpolation API"

This reverts commit 691a616e3e5473f97c6ad5001d5c38b97953dd81.

alexmoore committed:
Changing syntax around for TS 0.8, we wont have the Cell interpolation syntax until 0.9

alexmoore committed:
Moving tests to match ts 0.8 availability"

alexmoore committed:
Adding a few more edge case tests

alexmoore committed:
Switching to milliseconds for timestamps

alexmoore committed:
Cleaning up TS test duplication

alexmoore committed:
Reducing TS Test chattyness

alexmoore committed:
Fix Cell Test regressions introduced with millisecond changes

alexmoore committed:
Add test to show that timestamps come back in the Int buffer

alexmoore committed:
Fixing some test regressions, adding Cell docs

alexmoore committed:
Adding more notes about timestamps being returned

alexmoore committed:
[maven-release-plugin] prepare release riak-client-2.0.3-timeseries-beta-1

alexmoore committed:
[maven-release-plugin] prepare for next development iteration

alexmoore committed:
Removing unused columnDescription parameters from store commands

alexmoore committed:
Add test for null cells

alexmoore committed:
Adding support for real floats / doubles. More javadoc comments all around!

alexmoore committed:
Fix incomplete thought

alexmoore committed:
Adding tests around bad insert data / table names

alexmoore committed:
Document query float type & column description behavior

alexmoore committed:
Merge branch 'RTS-423-Floats' into RTS-405-Nulls

alexmoore committed:
Cleanup tests

alexmoore committed:
Merge branch 'timeseries' into RTS-405-Nulls

alexmoore committed:
Merge pull request #560 from basho/RTS-405-Nulls

Ready: Add Float/Doubles, and null support to Timeseries operations.

alexmoore committed:
Rename some files, move them to timeseries package/folder

alexmoore committed:
Fixing null bug, fixing timestamps coming back in long buffer, fixing bad rename refactoring

alexmoore committed:
Adding the PB builder to PBFutureOperation typespec, so we can add data to it after the super constructor runs.

alexmoore committed:
Work so far on single key reads. Blocked by need of Column Descriptions on TsGetResp

alexmoore committed:
First pass at Single delete command/operation

alexmoore committed:
Adding javadocs / cleanup

alexmoore committed:
Move TS op integration tests to new package folder.

alexmoore committed:
Integration tests for TS Delete command/operation

alexmoore committed:
Removing vClock from TS Delete, cannot get a vclock to use it currently

alexmoore committed:
Remove extra builder type from PBFutureOperation, do not really need it, makes everything ugly.

alexmoore committed:
Add fetch queryinfo implementation + unit tests.

alexmoore committed:
Test cleanup

alexmoore committed:
Tests for single key read

alexmoore committed:
Removing sets and maps

alexmoore committed:
Remove float32s :/

alexmoore committed:
Remove Int32 convenience methods, as they are misleading

alexmoore committed:
Remove unused column types

alexmoore committed:
Cleanup PBConverter Class

alexmoore committed:
More general cleanup

alexmoore committed:
More unused code

alexmoore committed:
Remove Numerics

alexmoore committed:
Merge pull request #561 from basho/RTS-483-Remove-Sets-Maps

Ready: Remove sets / maps / float32s from timeseries

alexmoore committed:
Change TS types to new names. Depends on basho/riak_pb#511

alexmoore committed:
Merge pull request #562 from basho/RTS-511-RemoveFloat

Change TS types to new names.

alexmoore committed:
Fixes for 3-part TS keys

alexmoore committed:
Merging latest changes in

alexmoore committed:
Merge branch 'fixes/am/3-part-ts-key' into RTS-449-TS-Single-Key-Delete

alexmoore committed:
Remove old merge files

alexmoore committed:
Fix delete test to match current TS implementation

alexmoore committed:
s/if(/if (/g

alexmoore committed:
PR fixes & suggestions

alexmoore committed:
Common response logging

alexmoore committed:
Reduce, reuse, refactor.

alexmoore committed:
PR suggestions

alexmoore committed:
remove some unneeded synchnization

alexmoore committed:
international pair programming

alexmoore committed:
Merge pull request #564 from basho/RTS-449-TS-Single-Key-Delete

Ready: Add single row fetches and deletes for TS. Move to 3-part key.

alexmoore committed:
Adding boolean column too, adding column types to single TS get results

alexmoore committed:
Renaming Binary to Varchar

alexmoore committed:
adding not found tests for TS single key read / delete

alexmoore committed:
Using different PB namespace than KVs

alexmoore committed:
Merge branch 'RiakTsPB-namespace' into fixes/am/binary_to_varchar

alexmoore committed:
Merge pull request #567 from basho/fixes/am/binary_to_varchar

Ready. Rename "binary" cells/columns to "varchar" to match recent changes.

alexmoore committed:
Take timestamps from timestamp register, not sint64 register

alexmoore committed:
Merge pull request #574 from basho/fixes/am/fix_decode_timestamp_cells

Take timestamps from timestamp register, not sint64 register

alexmoore committed:
Fix some out of date javadocs

alexmoore committed:
Merge branch 'timeseries' of https://github.com/basho/riak-java-client into timeseries

alexmoore committed:
Turn back on timeseries test check

alexmoore committed:
[maven-release-plugin] prepare release riak-client-2.0.3-timeseries-beta-2

alexmoore committed:
[maven-release-plugin] prepare for next development iteration

srgg committed:
Improve  TimeSeriesPBConverter.convertRowsToPb() performance

srgg committed:
Introduce QueryResult.EMPTY

srgg committed:
Improve and simplify  ColumnDescription.ColumnType conversion

srgg committed:
Getting rid of duplicated code in ITestQueryOperation

alexmoore committed:
Minor pedantic code formatting

srgg committed:
Introduces XLight TimeSeries API

alexmoore committed:
Merge pull request #576 from basho/tspb-encode-performance-improvements

TimeSeries Store Operation PB Encode Performance Improvements

srgg committed:
Simplify Riak cluster creation HostAndPort was added

alexmoore committed:
Add a few doc lines

alexmoore committed:
Boilerplate for TS List Keys

alexmoore committed:
Renaming a few classes, might have more than 1 immutable type in the future

alexmoore committed:
Use IQueryResult for ListKeys result, make ImmutablePbResultFactory / QueryResult more flexible for input

alexmoore committed:
Interface all the TS things!

srgg committed:
Reformat code

alexmoore committed:
Take out the interfaces, make the "Light" classes the new default TS classes.

alexmoore committed:
Rename & split ImmutablePbResultFactory, reformat & cleanup.

Rename & split ImmutablePbResultFactory into PbResultFactory and
CollectionConverters.

IntelliJ Reformat, & cleanup.

alexmoore committed:
Fix some test failures around Null cells

alexmoore committed:
Make ListKeysOp ctor private

srgg committed:
Simplifying RiakClient creation

srgg committed:
Reduce memory consumption in from/to PB Cell/Row conversions

alexmoore committed:
Introducing wrapped iterables, since PB looks at the iterator at least twice.

alexmoore committed:
Add list keys tests

srgg committed:
Code cleanup, minor renaming, getting rid of cached deep copies.

alexmoore committed:
Fix test

srgg committed:
Getting rid of BinaryValue in TS API.

alexmoore committed:
Merge branch 'ts-list-keys' of https://github.com/basho/riak-java-client into ts-list-keys

srgg committed:
Get rid of RiakNode.Builder.create()

alexmoore committed:
Merge pull request #577 from basho/feature/srgg/simplify-riak-cluster-creation

Simplify Riak cluster creation HostAndPort was added

alexmoore committed:
Implement SelectMany in Java

alexmoore committed:
Adjust tests for Eventual Consistency of ListKeys

alexmoore committed:
Merge pull request #579 from basho/ts-list-keys

Ready: TS List Keys + Lite/Immutable TS data classes.

alexmoore committed:
Adding a few missing javadocs, renaming a few parameters for clarity

alexmoore committed:
Merge branch 'develop' into timeseries

alexmoore committed:
Bump java version to 7

alexmoore committed:
Fix a few compiler warnings

alexmoore committed:
Make QueryResult implement Iterable<Row> explicitly

alexmoore committed:
Fix mojo pom bug and an invalid getDeclaredMethod() call

srgg committed:
Switch Delete, Store, Fetch commands as well as all the corespondig operations were changed to use Iterable.

alexmoore committed:
Brace placement cleanup, etc

srgg committed:
cleanup

srgg commented:
+1 after my last cleanup https://github.com/basho/riak-java-client/commit/a6e48fb24559b571946454da335fa5315cada4f0
